staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/matching/matching.go

```diff
@@ -0,0 +1,171 @@
+	m.namespaceMatcher.Client = client
+}
+
+func (m *Matcher) ValidateInitialization() error {
+	if err := m.namespaceMatcher.Validate(); err != nil {
+		return fmt.Errorf("namespaceMatcher is not properly setup: %v", err)
+	}
+	return nil
+}
+
+func (m *Matcher) Matches(criteria MatchCriteria, attr admission.Attributes, o admission.ObjectInterfaces) (bool, error) {
+	matches, err := m.namespaceMatcher.MatchNamespaceSelector(criteria, attr)
+	if err != nil {
+		return false, err
+	}
+	if !matches {
+		return false, nil
+	}
+
+	matches, err = m.objectMatcher.MatchObjectSelector(criteria, attr)
+	if err != nil {
+		return false, err
+	}
+	if !matches {
+		return false, nil
+	}
+
+	matchResources := criteria.GetMatchResources()
+	matchPolicy := matchResources.MatchPolicy
+	if matchesResourceRules(matchResources.ExcludeResourceRules, matchPolicy, attr, o) {
+		return false, nil
+	}
+
+	if !matchesResourceRules(matchResources.ResourceRules, matchPolicy, attr, o) {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func matchesResourceRules(namedRules []v1alpha1.NamedRuleWithOperations, matchPolicy *v1alpha1.MatchPolicyType, attr admission.Attributes, o admission.ObjectInterfaces) bool {
+	for _, namedRule := range namedRules {
+		rule := v1.RuleWithOperations(namedRule.RuleWithOperations)
+		ruleMatcher := rules.Matcher{
+			Rule: rule,
+			Attr: attr,
+		}
+		if !ruleMatcher.Matches() {
+			continue
+		}
+		// an empty name list always matches
+		if len(namedRule.ResourceNames) == 0 {
+			return true
+		}
+		// TODO: GetName() can return an empty string if the user is relying on
+		// the API server to generate the name... figure out what to do for this edge case
+		name := attr.GetName()
+		for _, matchedName := range namedRule.ResourceNames {
+			if name == matchedName {
+				return true
+			}
+		}
+	}
+
+	// if match policy is undefined or exact, don't perform fuzzy matching
+	// note that defaulting to fuzzy matching is set by the API
+	if matchPolicy == nil || *matchPolicy == v1alpha1.Exact {
+		return false
+	}
+
+	attrWithOverride := &attrWithResourceOverride{Attributes: attr}
+	equivalents := o.GetEquivalentResourceMapper().EquivalentResourcesFor(attr.GetResource(), attr.GetSubresource())
+	for _, namedRule := range namedRules {
+		for _, equivalent := range equivalents {
+			if equivalent == attr.GetResource() {
+				// we have already checked the original resource
+				continue
+			}
+			attrWithOverride.resource = equivalent
+			rule := v1.RuleWithOperations(namedRule.RuleWithOperations)
+			m := rules.Matcher{
+				Rule: rule,
+				Attr: attrWithOverride,
+			}
+			if !m.Matches() {
+				continue
+			}
+			// an empty name list always matches
+			if len(namedRule.ResourceNames) == 0 {
+				return true
+			}
+			// TODO: GetName() can return an empty string if the user is relying on
+			// the API server to generate the name... figure out what to do for this edge case
+			name := attr.GetName()
```

Comment:
another TODO about empty name

Reply:
Yes, It would be tracked with the generatedName(). My guess is we could use generatedName(if the generatedName happens earlier) but waiting for confirming from @aojea :)

Reply:
Update on this one: https://github.com/kubernetes/kubernetes/pull/112883#discussion_r1010920120 Looks like we just have to handle the scenario with empty name...